client/exec.go

```diff
@@ -0,0 +1,305 @@
+				return "", err
+			}
+			conns = append(conns, conn)
+			dones["stdin"] = wsutil.WebsocketSendStream(conn, args.Stdin, -1)
+			dones["stdout"] = wsutil.WebsocketRecvStream(args.Stdout, conn)
+		}
+
+		// Handle stderr separately if needed
+		if opts.Stderr && fds["stderr"] != "" {
+			conn, err := client.getChangeWebsocket(changeID, fds["stderr"])
+			if err != nil {
+				return "", err
+			}
+			conns = append(conns, conn)
+			dones["stderr"] = wsutil.WebsocketRecvStream(args.Stderr, conn)
+		}
+
+		// Wait for everything to be done
+		go func() {
+			for name, done := range dones {
+				// Skip stdin, dealing with it separately below
+				if name == "stdin" {
+					continue
+				}
+				<-done
+			}
+
+			if fds["io"] != "" {
+				if args.Stdin != nil {
+					args.Stdin.Close()
+				}
+
+				// Empty the stdin channel but don't block on it as
+				// stdin may be stuck in Read()
+				go func() {
+					<-dones["stdin"]
+				}()
+			}
+
+			for _, conn := range conns {
+				conn.Close()
+			}
+
+			if args.DataDone != nil {
+				close(args.DataDone)
+			}
+		}()
+	}
+
+	return changeID, nil
+}
+
+// getChangeWebsocket creates a websocket connection for the given change ID
+// and websocket ID combination.
+func (client *Client) getChangeWebsocket(changeID, websocketID string) (*websocket.Conn, error) {
+	// Set up a new websocket dialer based on the HTTP client
+	httpClient := client.doer.(*http.Client)
+	httpTransport := httpClient.Transport.(*http.Transport)
+	dialer := websocket.Dialer{
+		NetDial:          httpTransport.Dial,
+		Proxy:            httpTransport.Proxy,
+		TLSClientConfig:  httpTransport.TLSClientConfig,
+		HandshakeTimeout: 5 * time.Second,
+	}
+
+	// Establish the connection
+	url := fmt.Sprintf("ws://localhost/v1/changes/%s/websocket?id=%s", changeID, url.QueryEscape(websocketID))
+	conn, _, err := dialer.Dial(url, nil)
+	if err != nil {
+		return nil, err
+	}
+	return conn, err
+}
+
+// WebsocketWriter is a websocket writer interface that can write a websocket
+// or a value as JSON, for example, for sending commands to an executing
+// program's "control" websocket.
+type WebsocketWriter interface {
+	WriteMessage(messageType int, data []byte) error
+	WriteJSON(v interface{}) error
+}
+
+type execCommand struct {
+	Command string            `json:"command"`
+	Args    map[string]string `json:"args,omitempty"`
+	Signal  int               `json:"signal,omitempty"`
+}
+
+// ExecSendTermSize sends a window-resize message to the Exec control websocket.
+func ExecSendTermSize(conn WebsocketWriter, width, height int) error {
+	msg := execCommand{
+		Command: "window-resize",
+		Args: map[string]string{
+			"width":  strconv.Itoa(width),
+			"height": strconv.Itoa(height),
+		},
+	}
+	return conn.WriteJSON(msg)
+}
+
+// ExecForwardSignal forwards a signal to the Exec control websocket.
+func ExecForwardSignal(conn WebsocketWriter, signal int) error {
+	msg := execCommand{
+		Command: "signal",
+		Signal:  signal,
```

Comment:
I need to understand a bit more before making useful commands here, as I'm not even sure if this is something that is under our control or if we're just following a well established design, but my first feeling when reading this is that we could probably polish a bit the phrasing of these constructions.  For example, we have a general "args" field, but signal is something that is sent apart on its own field, although it's an argument to the command. Also, is a top-level function a better design for this case than a method on an opaque type that we define ourselves?  We don't really talk about the fact that this is being sent over a websocket in the client API for any other reason than the control messages, but looking at the exec command it doesn't look like there's much happening about websockets there.

But again, my view is shallow at this point, so please take these comments with a grain of salt. I need to read more.

Reply:
Yeah, this was a bit of a mess on both counts. My excuse is copying the design from LXD. :-) I made two cleanups:

* Made the command types/JSON more consistent and more type-safe. There's now a top-level `command` field, and then `*struct` fields (named with the command name) for each command with command-specific fields (each one is `omitempty` so only one set of args gets sent). I'm sure Haskell people would use sum types here, but this will do.
* I've also refactored so `Exec` returns an opaque `Execution` type, which has methods to `Wait` and send signals/resize messages. This has a few benefits:
    1) it's much simpler for the user
    2) it's closer to the Python API
    3) it avoids the messy `Control` goroutine which was (oddly) started by `Exec` itself (but didn't really have a way to stop)
    4) it moves the `DataDone` channel to an internal concern (though I may still change that per your comment above)

Reply:
Lots of good ideas there, and looks much better now, thanks!